python/dex/interop/jax/apply.py

```diff
@@ -110,11 +115,28 @@ def dex_apply_lowering(ctx, *args, func_atom):
   custom_call_name = get_trampoline()
   ctx.module_context.add_keepalive(native)
 
-  # TODO: Support inference of implicit arguments. Abstract eval already does inference!
-  if len(native.argument_signature) != len(native.explicit_argument_signature):
-    raise NotImplementedError("Inference of implicit arguments")
   assert len(args) == len(native.explicit_argument_signature)
 
+  # TODO Cache this from dex_call_abstract_eval_with_shape instead of
+  # recomputing here
+  shape_vars = unify_jax_and_dex_types(
+      ctx.avals_in, native.explicit_argument_signature)
+
+  name_to_mlir_arg = {}
+  for arg, binder in zip(args, native.explicit_argument_signature):
+    name_to_mlir_arg[binder.name] = arg
+
+  mlir_args = []
+  for binder in native.argument_signature:
+    if binder.implicit:
+      if binder.name in shape_vars:
+        mlir_arg = mlir.ir_constant(
+            IdxRepTy(shape_vars[binder.name]).value,
+            canonicalize_types=False)
+        mlir_args.append(mlir_arg)
+    else:
+      mlir_args.append(name_to_mlir_arg[binder.name])
```

Comment:
I feel like this is a bit roundabout. Could we basically do this?
```py
shape_vars = ...
name_to_mlir_arg = {}
for b, arg in zip(native.explicit_argument_signature, args):
  name_to_mlir_arg[b.name] = arg
for n, s in shape_vars.items():
  name_to_mlir_arg[n] = mlir.ir_constant(...)
mlir_args = [name_to_mlir_arg[n] for n in native.argument_signature]
```

I think that every implicit (shape) variable will always be an argument of the native function, so there's no loss in making that constants eagerly.

Reply:
Good idea!  Done.